src/averageAges.js

```diff
@@ -55,12 +84,35 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const peopleWithMother = people.filter(({ mother }) =>
+    mother !== null && people.find(({ name }) => name === mother));
+
+  const filteredPeopleWithMotherBySex = onlyWithSon
+    ? peopleWithMother.filter(({ sex }) => sex === SEX_MALE)
+    : peopleWithMother;
+
+  return filteredPeopleWithMotherBySex
+    .reduce((acc, child, index) => {
+      const { born, mother: motherName } = child;
+      const mother = people.find(person => person.name === motherName);
+
+      const ageDiff = mother
+        ? born - mother.born
+        : 0;
+
+      if (index === filteredPeopleWithMotherBySex.length - 1) {
+        return (acc + ageDiff) / filteredPeopleWithMotherBySex.length;
+      }
+
+      return acc + ageDiff;
+    }, 0);
 }
```

Comment:
The function efficiently calculates the average age difference, but consider simplifying the conditional check inside the reduce method for better readability.